[libc] [elkscmd] Simplify linking of programs

tkchia opened the pull request:
Hello all,

This series of patches simplifies the procedure for building ELKS userland programs (https://github.com/jbruchon/elks/issues/252).  The main changes are as follows:

  * The top-level `Makefile` now arranges to "install" `elks-libc` into the cross toolchain directory (under `cross/ia16-elf/lib/`).
  * The makefile rules for the ELKS userland utilities (`elkscmd/Make.defs`) can then link userland programs simply by saying `ia16-elf-gcc` ... `-melks-libc`, rather than having to manually pass `crt0.o`, `libc.a`, etc. to the linker.  This is because `ia16-elf-gcc` can now find the C library and include files at the expected places.

I further modified `make clean` at the top level so that it cleans away any "installed" `elks-libc` files in the cross toolchain directory.

To get everything to build (especially the `bc` utility), I also had to

  * add a simple `<stdint.h>` to the C library --- it just hands over to `libgcc`'s version;
  * add a function implementation of `<stdio.h>`'s `putchar( )`;
  * disable the `libgcc` routines in `elks-libc` (`libc/gcc/`) for now --- `ia16-elf-gcc` can now link programs with the actual `libgcc`, and having two versions of the `libgcc` functions sometimes causes linker errors.

Thank you!

tkchia committed:
[libc] add <stdint.h>, just forwards to libgcc's <stdint-gcc.h>

tkchia committed:
[libc] add an `uninstall' make target

tkchia committed:
[cmds] build utilities via ia16-elf-gcc driver rather than ld

This should pretty much complete the implementation needed
for https://github.com/jbruchon/elks/issues/252 .

tkchia committed:
[libc] Disable libgcc stuff in elks-libc

We can now easily link programs with the actual libgcc (as
setup by tools/build.sh).  Besides, having both elks-libc and
libgcc routines defined may sometimes result in linker errors.

tkchia committed:
[libc] add a putchar function implementation, to keep GCC happy

mfld-fr commented on libc/gcc/Makefile:
```diff
@@ -2,7 +2,10 @@
+# It is feasible to now link programs with the real libgcc, so this libgcc
+# module is not really needed any more.  -- tkchia 20190603
+# SRCS = divmodsi3.s ldivmod.s ashlsi3.s
+SRCS =
```
OK, but are the versions inside 'libgcc' as small as these ones (in term of machine code size) ?

tkchia replied:
Hello @mfld-fr,

The current `libgcc` implementations of these routines are [here](https://github.com/tkchia/gcc-ia16/blob/21b95e80adb7a080b850a9a047505ebc653ca580/libgcc/config/ia16/lib1funcs.S).  They are also coded in assembly language.  The `__ashlsi3` implementation is a little bit larger, but probably faster, than the version currently in the ELKS code.  I have not really closely compared the division routines yet.

Thank you!

tkchia replied:
Hello @mfld-fr,

Interestingly, the `libgcc` core routine for 32-bit division (`__ia16_ldivmodu`, taken from the FreeDOS kernel sources) is `0x4e` bytes --- less than half the size (!) of the `elks-libc` core routine (`ldivmod` + `ludivmod`), which is `0xda` bytes.  However, the `libgcc` implementation does not handle signed division yet.

I will try to implement signed 32-bit division in `libgcc` in assembly, and see how things go.

Thank you!

tkchia replied:
Hello @mfld-fr,

I have incorporated your `__ashlsi3` implementation into the `libgcc` code in the `-Os` case (see https://github.com/tkchia/gcc-ia16/commit/2dd73a5500dce4be8f0f7d4bab02f5ed6bf849b0), and further squished it --- so that it also limits the total shift count to `0x1f` but still takes up the same number of bytes.  (It is probably a bit slower now than the `elks-libc` implementation, though.)
```
00000000 <__ashlsi3>:
   0:	5b                   	pop    %bx
   1:	58                   	pop    %ax
   2:	5a                   	pop    %dx
   3:	59                   	pop    %cx
   4:	51                   	push   %cx
   5:	52                   	push   %dx
   6:	50                   	push   %ax
   7:	53                   	push   %bx
   8:	83 e1 1f             	and    $0x1f,%cx
   b:	e3 06                	jcxz   13 <__ashlsi3+0x13>
   d:	d1 e0                	shl    %ax
   f:	d1 d2                	rcl    %dx
  11:	e2 fa                	loop   d <__ashlsi3+0xd>
  13:	c3                   	ret    
```

I have also added signed 32-bit division routines `__divsi3` and `__modsi3` (https://github.com/tkchia/gcc-ia16/commit/55c23044b98b377a54b0aad50e9ee735b94a80b5), based on the existing unsigned division routine taken earlier from the FreeDOS kernel code.  The new implementation is shorter, and --- I think --- more conformant to C11, since it rounds quotients towards 0, as mandated by the standard.  (The `elks-libc` implementation in contrast rounds quotients towards −∞.)

I am still testing the changes and will let you know once this is done.

Thank you!

tkchia replied:
I have completed the tests and bumped up the GCC-IA16 version number.

mfld-fr commented on libc/Makefile:
```diff
@@ -65,7 +65,8 @@ SUBDIRS = \
-# $(PREFIX), such that they can be used from outside the ELKS source tree.
+# $(PREFIX), such that they can be used from outside the ELKS source tree;
+# and also add an `uninstall' rule to allow elks-libc to be cleaned away.
 #
```
PREFIX or DESTDIR according to Gnu standard ?

tkchia replied:
Hmm, yes, `DESTDIR` is probably a better choice here.  I will make the change on my end and update my `build-ia16` scripts correspondingly.

Thank you!

mfld-fr reviewed (commented):
Hello @tkchia, I would like to pull your changes, but I am still waiting for your response to my questions raised during the review. Could you please have a look to them ? Thanks !

tkchia committed:
[libc] `make install' into $(DESTDIR) rather than $(PREFIX)

Per suggestion by @mfld-fr
(https://github.com/jbruchon/elks/pull/272#discussion_r290228936).

tkchia committed:
[build] Update GCC-IA16 used version

This allows ELKS programs to be linked against the (IMO)
improved implementations of __ashlsi3, __lshrsi3, __divsi3, and
__modsi3 in the toolchain's libgcc.

tkchia commented:
Hello @mfld-fr,

> Hello @tkchia, I would like to pull your changes, but I am still waiting for your response to my questions raised during the review. Could you please have a look to them ? Thanks !

Thank you for the review!  I took the chance to incorporate improvements to `__ashlsi3`, `__divsi3`, and `__modsi3` into `libgcc` on my end.

Do let me know if the ELKS patches are now good to go.

Thank you!

mfld-fr commented:
All my questions answered. Thank you @tkchia !
